Try columns with grid and CSS vars

tellthemachines opened the pull request:
## Description
<!-- Please describe what you have changed or added -->

This is a proposal for drastically simplifying Columns/Column blocks styling with CSS grid and variables. 

The main reason for this is getting rid of the need to set inline styles in blocks as much as possible, because inline styles break the resizable editor (#19082).

This work will incidentally also fix #18416 and #19112 / #19515.

* Column width is now set in the columns block with CSS grid, but each column still computes its own width and may reset the width of the adjacent columns if total column widths exceed 100%.

* Columns without an explicit width are set to 1fr, so remaining width is distributed equally among all columns without explicit widths.

* I have added a basic flexbox fallback for IE, because this grid implementation depends on CSS variables to work. This means that IE users won't be able to preview columns with custom widths. 

* This setup is much easier to override by themes, because it doesn't depend as much on inline styles. E.g. themes can now cleanly override the grid-gap width with a single line of code. 


## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

Tested across browsers. IE maintains basic functionality.

## Screenshots <!-- if applicable -->

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->.

tellthemachines committed:
Try columns with grid and CSS vars

aristath commented:
hm.. It could also be as simple as 
```css
grid-template-columns: repeat(auto-fit, minmax(200px, 1fr));
```
That would simplify the code a lot and there would be no need for the css-var or calculating how many columns there are etc.

tellthemachines commented:
> hm.. It could also be as simple as
> 
> ```css
> grid-template-columns: repeat(auto-fit, minmax(200px, 1fr));
> ```

@aristath I'm afraid we can't use `auto-fit` or `auto-fill` here because we have to respect the both the user-specified number of columns and the custom column widths, if they are set.

tellthemachines committed:
Fix legacy columns and add IE fallback.

talldan commented:
This seems promising! I like the fact that unset columns mostly stay proportional.

In testing I had a few observations (tested using Twenty Nineteen):
- I selected the layout with two equal columns and then tried setting the first to 50%. I would've expected there to be no change, but when explicitly setting the value to 50% the first column seems to be larger than the second. I wonder if the gaps between columns needs to be taken into account in the calculation.
- Columns can expand beyond the block boundaries, particularly when there are 3 or more columns and one is set to a large size. I don't think this was possible before, or it wasn't as easy to achieve.
- When setting one column width a value set in other columns can become unset. This seems to be the case when the total adds up to over 100%. This was particularly noticeable the more columns there are. An example is with four columns, if I set the first to 90% and then accidentally drag the second to over 10%, the first suddenly snaps back to being small, which is unexpected.

draganescu commented on packages/block-library/src/columns/edit.js:
```diff
@@ -70,6 +73,37 @@ function ColumnsEditContainer( {
+		};
+	}, [ clientId ] );
+
+	let columnsTemplateString = ``;
```
This being the string that describes the grid, maybe it would be better-called something with 'grid' in it. Like 'gridValue' or 'gridStructure'? When I read template I think of the HTML of the columns or something ...

tellthemachines replied:
It's a CSS template 😁 
I'll switch it to the name of the CSS property we're actually using it on, `gridTemplateColumns`. That should make things clearer.

draganescu commented on packages/block-library/src/columns/editor.scss:
```diff
@@ -21,13 +21,34 @@
 		}
+
+		@supports (display: grid) {
+			--columns-template: 1fr;
```
same comment, might this be better named --grid-template ?

tellthemachines replied:
I think it's best to keep the explicit reference to columns here, because `grid-template` is an actual CSS property that combines `grid-template-columns` and `grid-template-rows`. I also didn't want the variable to have the exact name of the property because that might be confusing.

draganescu reviewed (commented):
This works good! 👏  Added just some minor comments.

draganescu commented:
> I selected the layout with two equal columns and then tried setting the first to 50%. I would've expected there to be no change, but when explicitly setting the value to 50% the first column seems to be larger than the second. I wonder if the gaps between columns needs to be taken into account in the calculation.

@talldan I tried this as well and the sizes remained the same. Although there is some flicker and i guess there remains an impression of the columns not being equal afterwards.

Also can we fix this weird state when the number next to the range control is empty?

<img width="1268" alt="Screenshot 2020-02-03 at 13 06 37" src="https://user-images.githubusercontent.com/107534/73648349-0adc6780-4686-11ea-9112-e963a8a01ee8.png">

tellthemachines commented on packages/block-library/src/column/edit.js:
```diff
@@ -55,6 +43,7 @@ function ColumnEdit( {
 					<RangeControl
 						label={ __( 'Percentage width' ) }
 						value={ width || '' }
+						initialPosition={ 0 }
```
This should be visible when a column has no explicit width set.

tellthemachines commented on packages/block-library/src/columns/edit.js:
```diff
@@ -70,6 +73,37 @@ function ColumnsEditContainer( {
+
+	childColumnsAttributes.forEach( ( column ) => {
+		if ( ! column.width || column.width < 0 ) {
+			columnsTemplateString += `1fr `;
```
Ugh, I don't think there's much we can do about that. Only legacy columns should have negative widths though; it's a bug from the previous implementation. The negative width will disappear as soon as the width for any of the columns in that block is updated. It's a bit of an edge case, and it won't happen with any new columns, so I didn't worry too much about it.

tellthemachines committed:
Switch percentage units with `fr`.

tellthemachines committed:
lint

tellthemachines committed:
Update snapshots.

tellthemachines commented on packages/block-library/src/column/edit.js:
```diff
@@ -113,44 +102,29 @@ export default compose(
-						columns.length
-					),
-				};
+				// Check if sum of all column widths exceeds 100.
```
It was, but in response to the issues @talldan raised with how the percentages were working, I have since changed to `fr` units. They are being used like percentage units though, so the `100` still makes sense here, and I added some (hopefully) explanatory comments.

tellthemachines commented:
@talldan I have updated this to use `fr` units instead of `%`, because `fr` take into account the `grid-gap` so there should be no more weirdness in the column sizing 😅 
I also changed the logic slightly to deal better with legacy column situations, where total sum of widths might exceed 100%. Exceeding 100% should no longer be possible with the changes in this PR.

talldan commented:
@tellthemachines Hmm, I seem to end up with really narrow columns when adding the block for some reason:
<img width="613" alt="Screen Shot 2020-02-06 at 11 26 26 am" src="https://user-images.githubusercontent.com/677833/73903297-af071e00-48d3-11ea-9d2b-c27023a5d014.png">

tellthemachines committed:
Adjust margins and grid-gaps

tellthemachines commented:
> Hmm, I seem to end up with really narrow columns when adding the block for some reason:

There were auto-margins on the inner columns. I've removed them and made some further margin-related tweaks. Everything should look as before now 😄

aduth commented:
I remarked at https://github.com/WordPress/gutenberg/issues/18416#issuecomment-593581168 with some concerns about reimplementing Columns using Grid.

tellthemachines commented:
Closing this PR as it will require a substantial re-write to fit with the recent changes to the columns block.
